packages/datx/src/helpers/model/utils.ts

```diff
@@ -257,10 +257,13 @@ export function assignModel<T extends PureModel>(model: T, key: string, value: a
       value instanceof PureModel;
 
     if (key in fields) {
-      if (shouldBeReference && !fields[key].referenceDef) {
+      const referenceDef = fields[key].referenceDef as IReferenceDefinition;
+      if (shouldBeReference && !referenceDef) {
         throw error('You should save this value as a reference.');
       }
-      model[key] = value;
+      if (!(referenceDef?.property && value === undefined)) {
```

Comment:
Why does it matter what the current value is?

Reply:
The specific reasons are as follows:
1. When adding a model to the collection, the `assignModel` function will evaluate the current value and update the model if it has an indirect reference, where the indirectly referenced value is' undefined '. Updating an indirect reference with 'undefined' throws an exception because the indirect reference is read-only. In fact, I didn't update the indirect reference, but `assignModel` calculates the indirect reference property, with the value 'undefined'. So this code is designed to filter for this.

2. I have a question: is `upsertModel({a,b})` equivalent to `upsertModel({a,b,c: undefined})` ? Or is `upsertModel({a,b})` equivalent to  `upsertModel({a,b, C :oldValue})` ?
Before the update process is started, all properties of the model will be recalculated. If the value of a property is calculated as' undefined ', what is the meaning of it?  delete the old value, or keep the old value? .

Reply:
1. makes sense
2. good question. Passing undefined is an information (compared to no value), so it might make sense to use it. I would say it should work just like `Object.assign`, so if you pass it undefined, it should set the value to undefined:
```
a = {a: 1, b: 2, c: 3}
> {a: 1, b: 2, c: 3}
Object.assign(a, {a: 4, c: undefined})
> {a: 4, b: 2, c: undefined}
```

Reply:
got it.

The following changes should be made

![image](https://user-images.githubusercontent.com/10877047/93456598-7e2ea300-f910-11ea-9cd8-2f3b703ea8e8.png)


The process of recalculation should be based on the `original data keys`  to filter `undefined` value, if there is a key in the `original data`, indicating the need to update, if not, indicating that this is a calculated undefined value, should be ignored.

Is that right?